src/Compiler/TypedTree/TypeProviders.fs

```diff
@@ -625,6 +630,36 @@ type ProvidedAssembly (x: Assembly) =
 
     override _.GetHashCode() = assert false; x.GetHashCode()
 
+[<NoComparison; NoEquality>]
+type CrackedStaticArgument = {
+    Name: string
+    Value: obj
+    ValueRange: range
+    /// Range stripped of quotes if the argument is a string constant
+    ValueRangeAdjusted: range option
+}
+
+type TypeProviderDiagnosticsContext (staticArgs: CrackedStaticArgument[], tpDesignation, fallbackRange) =
+    interface ITypeProviderDiagnosticsContext with
+        member _.ReportDiagnostic (staticParameterName, rangeInParameterIfString, message, severity) =
+            let m =
+                match staticArgs |> Array.tryFind (fun x -> x.Name = staticParameterName) with
+                | Some x when x.ValueRange <> range0 ->
+                    match x.ValueRangeAdjusted, rangeInParameterIfString with
+                    | Some m, Some (startPosition, endPosition) when startPosition < endPosition && startPosition >= 0 ->
+                        // todo highlighted range spanning multiple lines :/
+                        mkFileIndexRange m.FileIndex (mkPos m.StartLine (m.StartColumn + startPosition)) (mkPos m.EndLine (m.StartColumn + endPosition))
+                    | _ ->
+                        x.ValueRange
+                | _ ->
+                    fallbackRange
+
+            match severity with
+            | TypeProviderDiagnosticSeverity.Warning ->
+                Error (FSComp.SR.etProviderWarning (tpDesignation, message), m) |> warning
+            | TypeProviderDiagnosticSeverity.Error ->
+                stopProcessingRecovery (Error (FSComp.SR.etProviderError (tpDesignation, message), m)) m
```

Comment:
`stopProcessingRecovery` allows the TP to continue, but when I tried to use `Error`, which does take the control away, the diagnostic gets wrapped in additional exception, and the entire TP range gets a squiggly line, hiding what was meant to be highlighted. Not sure how to avoid that.